### PR TITLE
feat: Display advanced section within confirmation by default for some users

### PR DESCRIPTION
**Pull request by pedronfigueiredo**
## **Description**

Display advanced details section within confirmation by default for users that have edit nonce or showing hex data on. The PR includes e2e tests for these changes.

This PR also changes the copy for the nonce toggle and includes minor refactors to the gas fees logic.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/25687?quickstart=1)

## **Related issues**

Fixes: [#2737](https://github.com/MetaMask/MetaMask-planning/issues/2737)

## **Manual testing steps**

1. Enable nonce editing in the settings
2. Deploy an erc721 contract in the test dApp
3. Trigger an erc721 mint in the test dApp
4. The confirmation should come up with the advanced details toggled on

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/pages/confirmations/components/confirm/info/contexts/advanced-details-context.tsx
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/contexts/advanced-details-context.tsx:27]: I understood the final decision on this was to add a new property in the `PreferenceController` state to persist the advanced toggle preference.

And we'd do a one-off migration for anyone with custom name or hex data visible at the time, to set this new preference.

File: ui/pages/confirmations/components/confirm/info/contexts/advanced-details-context.tsx
Comment (OGPoyraz) [inline ui/pages/confirmations/components/confirm/info/contexts/advanced-details-context.tsx:21]: I think we better to add couple unit tests of this provider, since the main logic of this context provided there. 
Something like
```
it('initializes showAdvancedDetails based on Redux state', () => {
    getSendHexDataFeatureFlagStateMock.mockReturnValue(false);
    getUseNonceFieldMock.mockReturnValue(true);

    // Render the provider and a test consumer
    const { result } = renderHook(() => useAdvancedDetailsHandler(), {
      wrapper: AdvancedDetailsProvider,
    });

    // Assert that showAdvancedDetails is initialized correctly
    expect(result.current.showAdvancedDetails).toBe(true);
  });
  ```

File: app/scripts/controllers/preferences.js
Comment (jpuri) [inline app/scripts/controllers/preferences.js:97]: `isConfirmationAdvancedDetailsOpen` sounds more like app state then preference

File: ui/pages/confirmations/components/confirm/info/contexts/advanced-details-context.tsx
Comment (jpuri) [inline ui/pages/confirmations/components/confirm/info/contexts/advanced-details-context.tsx:28]: useMemo here does not looks very useful.

File: ui/pages/confirmations/components/confirm/info/contexts/advanced-details-context.test.tsx
Comment (jpuri) [inline ui/pages/confirmations/components/confirm/info/contexts/advanced-details-context.test.tsx:39]: `spyOn` may also be used here for mocking.

File: ui/components/app/confirm/info/row/text.tsx
Comment (jpuri) [inline ui/components/app/confirm/info/row/text.tsx:71]: I hope this left margin does not effect existing use cases.

File: test/e2e/tests/confirmations/transactions/contract-interaction-redesign.spec.js
Comment (jpuri) [inline test/e2e/tests/confirmations/transactions/contract-interaction-redesign.spec.js:494]: 👍

**Review by jpuri — DISMISSED** [review]
Great work 👍 
I left some small feedbacks

File: ui/components/app/confirm/info/row/text.tsx
Comment (pedronfigueiredo) [reply ui/components/app/confirm/info/row/text.tsx:71]: I made the change in the name of consistency with the designs, so I think it's safe.

File: ui/pages/confirmations/components/confirm/info/contexts/advanced-details-context.tsx
Comment (pedronfigueiredo) [reply ui/pages/confirmations/components/confirm/info/contexts/advanced-details-context.tsx:28]: The `useMemo` is primarily used for maintaining referential equality and potentially preventing unnecessary re-renders in child components.

File: app/scripts/controllers/preferences.js
Comment (pedronfigueiredo) [reply app/scripts/controllers/preferences.js:97]: It is an app state though, we substituted `useState` with this background preference property for expanding and collapsing advanced details...

File: app/scripts/migrations/122.ts
Comment (matthewwalsh0) [inline app/scripts/migrations/122.ts:41]: As this is the persisted state and not the UI state, should these be `state?.PreferencesController?.useNonceField`?

File: app/scripts/migrations/122.ts
Comment (matthewwalsh0) [inline app/scripts/migrations/122.ts:35]: If we need it below, could we assign this and check it with a standard falsey check?

```
const preferencesState = state?.PreferencesController;

if(!preferencesState) {
    return state;
}
```

File: app/scripts/migrations/122.ts
Comment (matthewwalsh0) [inline app/scripts/migrations/122.ts:47]: Could we simplify to:

```
state.PreferencesController.preferences.isConfirmationAdvancedDetailsOpen = isCustomNonceFieldEnabled || isHexDataVisibilityEnabled;
```

File: app/scripts/controllers/preferences.js
Comment (matthewwalsh0) [reply app/scripts/controllers/preferences.js:97]: Even though it's triggered via a toggle, it is still a mechanism to indicate their preference which we could theoretically use elsewhere.

So maybe `showConfirmationAdvancedDetails`?

File: ui/pages/confirmations/components/confirm/info/contexts/advanced-details-context.tsx
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/contexts/advanced-details-context.tsx:11]: As this is now just reliant on standard controller state, is it fair to say a React context introduces unnecessary complexity?

We can simply use the selector where we need to read the value, and a dispatch where we want to set it as we do with all controller state?

File: ui/selectors/selectors.js
Comment (matthewwalsh0) [inline ui/selectors/selectors.js:2532]: Can we avoid adding selectors to this legacy huge file and instead use our more targeted `ui/pages/confirmations/selectors/confirm.ts`?

Comment (matthewwalsh0) [inline ui/selectors/selectors.js:2532]: For clarity, I think we're trying to name all new selectors `selectX` as recommended by Redux to make it explicit what are selectors and what are utils etc?

Maybe `selectShowConfirmationsAdvancedDetails`?

File: ui/selectors/selectors.js
Comment (matthewwalsh0) [inline ui/selectors/selectors.js:2533]: Minor, but maybe we can avoid the double reference with:

```
return Boolean(getPreferences(state).isConfirmationAdvancedDetailsOpen);
```

**Review by matthewwalsh0 — CHANGES_REQUESTED** [review]

File: ui/pages/confirmations/components/confirm/info/contexts/advanced-details-context.tsx
Comment (matthewwalsh0) [reply ui/pages/confirmations/components/confirm/info/contexts/advanced-details-context.tsx:28]: I think the container for the properties typically shouldn't matter, as any dependent component or hook would be using the values within it the object as the dependencies, rather than the container itself.

File: app/scripts/migrations/122.ts
Comment (matthewwalsh0) [inline app/scripts/migrations/122.ts:44]: Can we remove the `return` since we don't use it above?

File: app/scripts/migrations/122.ts
Comment (matthewwalsh0) [inline app/scripts/migrations/122.ts:36]: Minor, but we could have a `preferencesControllerState` and a `preferences` variable derived from it?

File: app/scripts/migrations/122.ts
Comment (matthewwalsh0) [inline app/scripts/migrations/122.ts:34]: Minor, but could we early return to avoid nesting the core logic?

File: test/e2e/tests/metrics/errors.spec.js
Comment (matthewwalsh0) [inline test/e2e/tests/metrics/errors.spec.js:860]: Intentional?

**Review by matthewwalsh0 — DISMISSED** [review]

File: test/e2e/tests/metrics/errors.spec.js
Comment (pedronfigueiredo) [reply test/e2e/tests/metrics/errors.spec.js]: removed

File: app/scripts/migrations/122.ts
Comment (pedronfigueiredo) [reply app/scripts/migrations/122.ts]: I also prefer to return early but this sonarqube quality gate blocks the PR when using that pattern.

https://sonarcloud.io/organizations/metamask/rules?open=typescript%3AS3516&rule_key=typescript%3AS3516

File: test/e2e/tests/confirmations/transactions/contract-interaction-redesign.spec.ts
Comment (seaona) [inline test/e2e/tests/confirmations/transactions/contract-interaction-redesign.spec.ts:165]: This spec has several points that need to be changed, as a first step of fixing the tests:
- we should leverage fixtures for having the nonce enabled, instead of going to settings
```
.withPreferencesController({
      preferences: { redesignedConfirmationsEnabled: true },
      useNonceField: true, // enable nonce here
})
```
- we should remove the step of going to settings and enabling nonce
- we should remove the step of deploying a contract (it's already deployed)
- we should remove the step of confirming the contract deployment (it's already deployed)


```
// all 3 steps should be removed
await toggleOnCustomNonce(driver);
await createContractDeploymentTransaction(driver);
await confirmContractDeploymentTransaction(driver);
```

- we should login with balance validation to mitigate the flakiness point where the dialog does not open, as the wallet is not ready  `await logInWithBalanceValidation(driver);`
- we can add the ` await driver.waitUntilXWindowHandles(3);` before switching to the dialog window

Once this is done, we see that the test now fails consistently on the edit nonce step. This is actually a legit failure as I don't see the nonce field in the transaction screen, when we have the redesign toggle enabled, so it's something to fix on the app level.

https://github.com/user-attachments/assets/43aef96c-17dd-4233-83f3-3299c1394e32




Here is the e2e with the changes implemented and the current legit failure:

https://github.com/user-attachments/assets/33915b1d-848e-44e2-95d6-b78665a60f03

**Review by matthewwalsh0 — APPROVED** [review]

**Review by jpuri — APPROVED** [review]

**Pull request merged**
